statsmodels/tsa/innovations/_arma_innovations.pyx.in

```diff
@@ -0,0 +1,734 @@
+    "z": ("np.complex128_t", "complex", "np.NPY_COMPLEX128"),
+}
+
+}}
+
+# Typical imports
+import numpy as np
+import warnings
+from statsmodels.tsa import arima_process
+from statsmodels.src.math cimport *
+cimport numpy as np
+cimport cython
+
+np.import_array()
+
+cdef int C = 0
+
+
+{{for prefix, types in TYPES.items()}}
+{{py:cython_type, dtype, typenum = types}}
+{{py:
+combined_prefix = prefix
+combined_cython_type = cython_type
+if prefix == 'c':
+    combined_prefix = 'z'
+    combined_cython_type = 'np.complex128_t'
+if prefix == 's':
+    combined_prefix = 'd'
+    combined_cython_type = 'np.float64_t'
+}}
+
+
+cdef {{prefix}}toeplitz(int n, int o0, int o1,
+                        {{cython_type}} [:] in_column,
+                        {{cython_type}} [:, :] out_matrix):
+    """
+    toeplitz(int n, int o0, int o1, {{cython_type}} [:] in_column, {{cython_type}} [:, :] out_matrix
+    
+    Construct a Toeplitz block in a matrix in place
+
+    Parameters
+    ----------
+    n : int
+        The number of entries of `in_column` to use.
+    o0 : int
+        The row offset for `out_matrix` at which to begin writing the Toeplitz
+        block.
+    o1 : int
+        The column offset for `out_matrix` at which to begin writing the
+        Toeplitz block.
+    in_column : ndarray
+        The column used to construct the Toeplitz block.
+    out_matrix : ndarray
+        The matrix in which to write a Toeplitz block
+
+    Notes
+    -----
+    This function constructs the Toeplitz block in-place and does not return
+    any output.
+
+    """
+    cdef int i, j
+
+    for i in range(n):
+        for j in range(i + 1):
+            out_matrix[o0 + i, o1 + j] = in_column[i - j]
+            if i != j:
+              # Note: scipy by default does complex conjugate, but not
+              # necessary here since we're only dealing with covariances,
+              # which will be real
+              out_matrix[o0 + j, o1 + i] = in_column[i - j]
+
+
+def {{prefix}}arma_transformed_acovf({{cython_type}} [:] ar,
+                                    {{cython_type}} [:] ma,
+                                    {{cython_type}} [:] arma_acovf):
+    """
+    arma_transformed_acovf({{cython_type}} [:] ar, {{cython_type}} [:] ma, {{cython_type}} [:] arma_acovf)
+    
+    Construct the autocovariance matrix for a transformed process.
+
+    Using the autocovariance function for an ARMA process, constructs the
+    autocovariance matrix associated with the transformed process described
+    in equation (3.3.1) of _[1].
+
+    Parameters
+    ----------
+    ar : ndarray
+        Autoregressive coefficients, including the zero lag, where the sign
+        convention assumes the coefficients are part of the lag polynomial on
+        the left-hand-side of the ARMA definition (i.e. they have the opposite
+        sign from the usual econometrics convention in which the coefficients
+        are on the right-hand-side of the ARMA definition).
```

Comment:
I think this is a good idea. Have we already started to do this anywhere, or do we need to establish a new convention?

@josef-pkt ? @bashtage ?

Reply:
I see that in `SARIMAX` I have `SARIMAXResults.polynomial_reduced_ar` that gives e.g. `[1, -phi_1, -phi_2, ...] and `SARIMAXResults.polynomial_reduced_ma` that gives e.g. `[1, theta_1, theta_2, ...]`.

where the model is defined as:

![equation](http://www.sciweavers.org/tex2img.php?eq=%5Cphi%28L%29%20y_t%20%3D%20%5Ctheta%28L%29%20%5Cvarepsilon_t&bc=White&fc=Black&im=jpg&fs=12&ff=arev&edit=)

Reply:
Well I guess that model wasn't too helpful. What I mean is that the model is:

![equation](http://www.sciweavers.org/tex2img.php?eq=y_t%20%3D%20%5Csum_%7Bi%3D1%7D%5Ep%20%5Cphi_i%20y_%7Bt-i%7D%20%2B%20%5Csum_%7Bi%3D1%7D%5Eq%20%5Ctheta_i%20%5Cvarepsilon_%7Bt-q%7D%20%2B%20%5Cvarepsilon%20&bc=White&fc=Black&im=jpg&fs=12&ff=arev&edit=)

and then the lag polynomials are:

- ![equation](http://www.sciweavers.org/tex2img.php?eq=%5Cphi%28L%29%20%3D%201%20-%20%5Cphi_1%20L%20-%20%5Cdots%20-%20%5Cphi_p%20L%5Ep&bc=White&fc=Black&im=jpg&fs=12&ff=arev&edit=)
- ![equation](http://www.sciweavers.org/tex2img.php?eq=%5Ctheta%28L%29%20%3D%201%20%20%2B%20%5Ctheta_1%20L%20%2B%20%5Cdots%20%2B%20%5Ctheta_q%20L%5Eq&bc=White&fc=Black&im=jpg&fs=12&ff=arev&edit=)

so e.g. `SARIMAXReuslts.polynomial_reduced_ar` gives the coefficients of \phi(L)

Reply:
so what about one of (listed in order of my preference):

- `ar_polynomial`
- `arpoly`
- `polynomial_ar`

Reply:
Very much +1 for establishing a standard with unambiguous names and using it consistently to clear up `(ar|ma)(params|coefs|poly)`.  Slight preference for `arpoly` over `ar_polynomial`.

Reply:
on the other hand, we already have `ar_params`, so what about `ar_poly`?

Reply:
(mainly checking if my new internet connection works)

I thought we are reasonably consistent

`params`, `ar_coeffs` are econometrics estimation version
`ar` is lag polynomial coefficients
ar with `poly` in the name should be a numpy Polynomial or similar instance which can be used to directly compute with the polynomials, e.g. multiplication, inverse and so on.

Reply:
> I thought we are reasonably consistent

I'd speculate that past-me would may have found `ar` and `ma` ambiguous, possibly been confused why there are both `arparams` and `arcoeffs` if they mean the same thing.  I also wouldn't be surprised if there are test/example/sandbox/pseudo-sandbox files that don't follow the same conventions.

Quick look grepping for `(ar|ma)_?(params|coef)` suggests that code looks appreciably better than the tests consistency/clarity-wise:

`ArmaProcess.__init__` takes `ar` and `ma`
`ARMA.geterrors` defines `b, a = np.r_[1, -arparams], np.r_[1, maparams]`
`arma_mle` does its own thing
`varma_process.VarmaPoly.__init__` takes `ar` and `ma`, looks vaguely like it follows the convention.

`test_arima.test_long_ar_start_params` defines `arparams` and `maparams` where it should be `ar` and `ma`.
`test_arima.test_arima_predict_exog` has commented-out code defining `arparams` and `maparams` with leading 1s, looks inconsistent, but who knows with commented-out code.
`test_arima.test_arima_1123` passes no-leading-1s `arparams`/`maparams` to `arma_generate_sample`, which AFAICT doesn't match the intended usage.

`test_stattools.test_arma_order_select_ic` starts out strong before taking its eyes off the prize:
```
    arparams = np.array([.75, -.25])
    maparams = np.array([.65, .35])
    arparams = np.r_[1, -arparams]
    maparam = np.r_[1, maparams]
```

Not going to bother looking at sandbox.

In no particular order, I'd suggest:
- fixing (or clarifying if my reading is incorrect) the apparently-inconsistent usages in the test files (side-note: delete-or-document commented-out code)
 -  adding validation in `arma_generate_sample` if `test_arima_1123` is in fact passing invalid params (side-note: give test_arima_1123 a meaningful name)
- using something more explicit than `ar`/`ma`, e.g. `arpoly`/`mapoly`
  - could consider going even further with this and only allow np.polynomial.Polynomial objects for these
- settling on either `(ar|ma)params` or `(ar|ma)coeffs`, but not using both synonymously.
- have ArmaProcess and ARMAResults share code to decrease the potential for future inconsistencies

Reply:
tl;dr - it sounds like a precedent already exists to use `ar` for the polynomial and `ar_params` for the econometric versions, so that's what I'll do here too.

> I thought we are reasonably consistent
> 
> params, ar_coeffs are econometrics estimation version
ar is lag polynomial coefficients
ar with poly in the name should be a numpy Polynomial or similar instance which can be used to directly compute with the polynomials, e.g. multiplication, inverse and so on.

Thanks, I had forgotten about the `ar` vs `ar_params` distinction. I think you're right, that this is reasonably consistent in statsmodels as of now.

> Quick look grepping for (ar|ma)_?(params|coef) suggests that code looks appreciably better than the tests consistency/clarity-wise:

Thanks for taking a look into the consistency in the existing code.

> In no particular order, I'd suggest:

I support those goals, but I guess out of scope for this PR.